File: 3435.frequencies-of-shortest-supersequences.go
#
# @lc app=leetcode id=3435 lang=golang
#
# [3435] Frequencies of Shortest Supersequences
#

# @lc code=start
func supersequences(words []string) [][]int {
	charSet := make(map[byte]bool)
	for _, word := range words {
		charSet[word[0]] = true
		charSet[word[1]] = true
	}

	letterList := []int{}
	var idOf [26]int
	for i := range idOf {
		idOf[i] = -1
	}

	idIdx := 0
	for ch := range charSet {
		letIdx := int(ch - 'a')
		letterList = append(letterList, letIdx)
		idOf[letIdx] = idIdx
		idIdx++
	}

	k := len(letterList)

	mand := 0
	pairs := [][2]int{}
	for _, word := range words {
		u := idOf[int(word[0]-'a')]
		v := idOf[int(word[1]-'a')]
		pairs = append(pairs, [2]int{u, v})
		if u == v {
			mand |= 1 << u
		}
	}

	minL := 999
	var res [][]int

	for mask := 0; mask < (1 << k); mask++ {
		if (mask & mand) != mand {
			continue
		}

		var openID [16]int
		var closeID [16]int
		eid := 0
		for i := 0; i < k; i++ {
			openID[i] = eid
			eid++
			if (mask & (1 << i)) != 0 {
				closeID[i] = eid
				eid++
			} else {
				closeID[i] = openID[i]
			}
		}

		E := eid

		adjj := make([][]int, E)
		indegg := make([]int, E)

		// double edges
		for i := 0; i < k; i++ {
			if (mask & (1 << i)) != 0 {
				fr := openID[i]
				to := closeID[i]
				adjj[fr] = append(adjj[fr], to)
				indegg[to]++
			}
		}

		// pair edges
		for _, p := range pairs {
			u := p[0]
			v := p[1]
			fr := openID[u]
			to := closeID[v]
			adjj[fr] = append(adjj[fr], to)
			indegg[to]++
		}

		// Kahn's
		q := []int{}
		for i := 0; i < E; i++ {
			if indegg[i] == 0 {
				q = append(q, i)
			}
		}

		qi := 0
		cnt := 0
		for qi < len(q) {
			u := q[qi]
			qi++
			cnt++
			for _, nei := range adjj[u] {
				indegg[nei]--
				if indegg[nei] == 0 {
					q = append(q, nei)
				}
			}
		}

		ok := (cnt == E)

		if ok {
			thisL := E
			if thisL < minL {
				minL = thisL
				res = res[:0]
				freq := make([]int, 26)
				for i := 0; i < k; i++ {
					fcnt := 1
					if (mask & (1 << i)) != 0 {
						fcnt = 2
					}
					let := letterList[i]
					freq[let] = fcnt
				}
				res = append(res, freq)
			} else if thisL == minL {
				freq := make([]int, 26)
				for i := 0; i < k; i++ {
					fcnt := 1
					if (mask & (1 << i)) != 0 {
						fcnt = 2
					}
					let := letterList[i]
					freq[let] = fcnt
				}
				res = append(res, freq)
			}
		}
	}

	return res
}
# @lc code=end

File: 3435.frequencies-of-shortest-supersequences.java
#
# @lc app=leetcode id=3435 lang=java
#
# [3435] Frequencies of Shortest Supersequences
#

# @lc code=start
import java.util.*;
class Solution {
  public List<List<Integer>> supersequences(String[] words) {
    Set<Character> charSet = new HashSet<>();
    for (String w : words) {
      charSet.add(w.charAt(0));
      charSet.add(w.charAt(1));
    }
    List<Character> letters = new ArrayList<>(charSet);
    int k = letters.size();
    if (k == 0) return new ArrayList<>();
    Map<Character, Integer> idmap = new HashMap<>();
    for (int i = 0; i < k; i++) {
      idmap.put(letters.get(i), i);
    }
    List<int[]> edges = new ArrayList<>();
    for (String w : words) {
      int u = idmap.get(w.charAt(0));
      int v = idmap.get(w.charAt(1));
      edges.add(new int[]{u, v});
    }
    int min_r = Integer.MAX_VALUE;
    for (int dmask = 0; dmask < (1 << k); dmask++) {
      if (isValid(dmask, k, edges)) {
        min_r = Math.min(min_r, Integer.bitCount(dmask));
      }
    }
    List<List<Integer>> ans = new ArrayList<>();
    for (int dmask = 0; dmask < (1 << k); dmask++) {
      if (Integer.bitCount(dmask) == min_r && isValid(dmask, k, edges)) {
        List<Integer> freq = new ArrayList<>();
        for (int i = 0; i < 26; i++) freq.add(0);
        for (int x = 0; x < k; x++) {
          int cnt = ((dmask & (1 << x)) != 0) ? 2 : 1;
          char ch = letters.get(x);
          freq.set(ch - 'a', cnt);
        }
        ans.add(freq);
      }
    }
    return ans;
  }
  private boolean isValid(int dmask, int k, List<int[]> edges) {
    int[] node_first = new int[k];
    int[] node_last = new int[k];
    int node_cnt = 0;
    for (int x = 0; x < k; x++) {
      node_first[x] = node_cnt++;
      node_last[x] = ((dmask & (1 << x)) != 0) ? node_cnt++ : node_first[x];
    }
    List<List<Integer>> g = new ArrayList<>();
    for (int i = 0; i < node_cnt; i++) {
      g.add(new ArrayList<>());
    }
    int[] indegree = new int[node_cnt];
    for (int x = 0; x < k; x++) {
      if ((dmask & (1 << x)) != 0) {
        int fr = node_first[x];
        int to = node_last[x];
        g.get(fr).add(to);
        indegree[to]++;
      }
    }
    for (int[] e : edges) {
      int fr = node_first[e[0]];
      int to = node_last[e[1]];
      g.get(fr).add(to);
      indegree[to]++;
    }
    Queue<Integer> q = new LinkedList<>();
    int processed = 0;
    for (int i = 0; i < node_cnt; i++) {
      if (indegree[i] == 0) q.offer(i);
    }
    while (!q.isEmpty()) {
      int u = q.poll();
      processed++;
      for (int nei : g.get(u)) {
        indegree[nei]--;
        if (indegree[nei] == 0) {
          q.offer(nei);
        }
      }
    }
    return processed == node_cnt;
  }
}
# @lc code=end